include/cantera/numerics/AdaptivePreconditioner.h

```diff
@@ -0,0 +1,300 @@
+/**
+ *  @file AdaptivePreconditioner.h
+ *   Declarations for the class AdaptivePreconditioner
+ *   which is a child class of PreconditionerBase for preconditioners used by sundials
+ */
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#ifndef ADAPTIVEPRECONDITIONER_H
+#define ADAPTIVEPRECONDITIONER_H
+
+#if CT_USE_SYSTEM_EIGEN
+#include <Eigen/Sparse>
+#else
+#include "cantera/ext/Eigen/Sparse"
+#endif
+
+#include "cantera/numerics/PreconditionerBase.h"
+#include "cantera/kinetics/ReactionDerivativeManager.h"
+#include "cantera/zerodim.h"
+#include "cantera/thermo.h"
+#include "cantera/kinetics.h"
+#include "cantera/kinetics/Reaction.h"
+
+#include "float.h"
+#include <unordered_map>
+
+namespace Cantera
+{
+
+//! Flag to indicate adaptive preconditioner is set
+const int ADAPTIVE_MECHANISM_PRECON_MATRIX = 1;
+
+class AdaptivePreconditioner : public PreconditionerBase
+{
+    protected:
+        //! @param m_values is the container for the values that are mapped to m_jacobian
+        std::vector<double> m_values;
+
+        //! @param m_inner is the container for inner (row) indices in CSC format
+        std::vector<int> m_inner;
```

Comment:
For brevity, there are some widely used typedefs for this (from `ct_defs.h`, which is explicitly or implicitly included just about everywhere):
```suggestion
        vector_fp m_values;

        //! @param m_inner is the container for inner (row) indices in CSC format
        vector_int m_inner;
```